File: Python/307__Range_Sum_Query-Mutable.py
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.nums = nums
        self.add = nums[::]
        for i in range(1, len(nums)):
            self.add[i] += self.add[i - 1]

    def update(self, i, val):
        """
        :type i: int
        :type val: int
        :rtype: void
        """
        diff = val - self.nums[i]
        self.nums[i] = val
        for index in range(i, len(self.nums)):
            self.add[index] += diff


    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        if i == 0:
            return self.add[j]
        return self.add[j] - self.add[i - 1]



# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(i,val)
# param_2 = obj.sumRange(i,j)

File: Java/307__Range_Sum_Query-Mutable.java
class NumArray {

    int[] tree;
    int[] nums;
    int size;

    public NumArray(int[] nums) {
        this.size = nums.length;
        this.nums = new int[size];
        this.tree = new int[size];
        for (int i = 0; i < size; i++)
            update(i, nums[i]);
    }

    public void update(int i, int val) {
        int delta = val - nums[i];
        nums[i] = val;
        for (; i < size; i |= i + 1)
            tree[i] += delta;
    }

    public int sumRange(int i, int j) {
        return sum(j) - sum(i - 1);
    }

    public int sum(int ind) {
        int result = 0;
        while (ind >= 0) {
            result += tree[ind];
            ind &= ind + 1;
            ind--;
        }
        return result;
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(i,val);
 * int param_2 = obj.sumRange(i,j);
 */
